src/main/java/com/bootcamp/demo/restapi/ScooterController.java

```diff
@@ -0,0 +1,58 @@
+package com.bootcamp.demo.restapi;
+
+import com.bootcamp.demo.model.Scooter;
+import com.bootcamp.demo.service.ScooterService;
+import com.bootcamp.demo.service.exception.ServiceException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+import org.springframework.http.ResponseEntity;
+import org.springframework.web.bind.annotation.*;
+
+import java.util.Set;
+
+@RestController
+@RequestMapping(path = "/api/scooters")
+public class ScooterController {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(ScooterController.class);
+    private final ScooterService scooterService;
+
+    public ScooterController(ScooterService scooterService) {
+        this.scooterService = scooterService;
+    }
+
+    @GetMapping
+    @ResponseBody
+    public ResponseEntity<Set<Scooter>> getScooters() {
+        Set<Scooter> scooters = null;
+        try {
+            scooters = scooterService.findAllScooters();
+            return new ResponseEntity<>(scooters, HttpStatus.OK);
+        } catch (ServiceException e) {
+            return new ResponseEntity<>(scooters, HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    @GetMapping("/{scooterId}")
+    @ResponseBody
+    public ResponseEntity<Object> getScooterById(@PathVariable(value = "scooterId") String scooterId) {
+        try {
+            Scooter scooter = scooterService.findScooterById(scooterId);
+            return new ResponseEntity<>(scooter, HttpStatus.OK);
+        } catch (ServiceException e) {
+            return new ResponseEntity<>(HttpStatus.BAD_REQUEST);
+        }
+    }
+
+    @PostMapping
+    public ResponseEntity<Object> postScooter(@RequestBody final Scooter scooter) {
+        try {
+            scooter.getBattery().setStatus();
```

Comment:
Could you outline why it is necessary to call this directly here? Note the following:
- If empty constructor is used for deserialization, then the fields from `Battery` will be set through setters. In this case, the `setLevel` also calls the `setStatus` method.
- If the constructor with all params is used for deserialization, then the `setStatus` method is called in the constructor.
Thus, in both cases, the Battery should be created in a consistent manner & have status set appropriately.
If that is not the case, could you provide details regarding what happens?

Reply:
And have you looked into why this happens (through debugging)?
From what you have outlined, it looks like it is caused by the fact that `setLevel` (which would set the status to LOW) is called before `setStatus` (which will set the status to the incorrectly provided value). Although having afterwards the setter called here explicitly solves the problem for now, it can lead to subtle future issues & introduces an unwanted coupling between creation & always calling this method afterwards.
What do you think about the following as a solution to not need this: Have the `Battery` class with the 2 fields that it has now, but not expose serialise status for it? This would result in having the following serialised structure (as json):
```
{
  "serialNumber": "string",
  "currentLocation": {
    "latitude": 0,
    "longitude": 0
  },
  "battery": {
    "level": 0
  },
  "status": "AVAILABLE"
}
```
but when operating with the model in the app, we would still have the status (and can be used when needed).
This can be easily achieved by annotating the `getStatus` with `@JsonIgnore`.

Reply:
That solved partially the problem, because when I make a GET request then the battery status is not included in the response, which is not what I wanted tho. (its ignored in both direction).

Reply:
Indeed, it is not shown, but wondering if it is shown/needed currently. If it is, we could have this limitation/coupling for the moment, but we should probably document it, so that for possible further updates for a scooter, this is taken into account.